rust/sdk-processor/src/steps/events_processor/events_extractor.rs

```diff
@@ -0,0 +1,81 @@
+use crate::db::common::models::events_models::events::EventModel;
+use aptos_indexer_processor_sdk::{
+    aptos_protos::transaction::v1::{transaction::TxnData, Transaction},
+    steps::{async_step::AsyncRunType, AsyncStep},
+    traits::{NamedStep, Processable},
+    types::transaction_context::TransactionContext,
+    utils::errors::ProcessorError,
+};
+use async_trait::async_trait;
+use rayon::prelude::*;
+use tracing::warn;
+
+pub const MIN_TRANSACTIONS_PER_RAYON_JOB: usize = 64;
+
+pub struct EventsExtractor
+where
+    Self: Sized + Send + 'static, {}
+
+#[async_trait]
+impl Processable for EventsExtractor {
+    type Input = Transaction;
+    type Output = EventModel;
+    type RunType = AsyncRunType;
+
+    async fn process(
+        &mut self,
+        item: TransactionContext<Transaction>,
+    ) -> Result<Option<TransactionContext<EventModel>>, ProcessorError> {
+        let events = item
+            .data
+            .par_iter()
+            .with_min_len(MIN_TRANSACTIONS_PER_RAYON_JOB)
+            .map(|txn| {
+                let mut events = vec![];
+                let txn_version = txn.version as i64;
+                let block_height = txn.block_height as i64;
+                let txn_data = match txn.txn_data.as_ref() {
+                    Some(data) => data,
+                    None => {
+                        warn!(
```

Comment:
When do we expect this to happen? Should this be error! or even panic!?

Reply:
This happens when processing block epilogue transactions, it doesn't have transaction data.